jupyter_server/extension/manager.py

```diff
@@ -97,6 +129,29 @@ def module(self):
         """The imported module (using importlib.import_module)"""
         return self._module
 
+    @property
+    def tools(self):
+        """Structured tools exposed by this extension point, if any."""
+        loc = self.app or self.module
+        if not loc:
+            return {}
+
+        tools_func = getattr(loc, "jupyter_server_extension_tools", None)
+        if not callable(tools_func):
+            return {}
+
+        tools = {}
+        try:
+            definitions = tools_func()
+            for name, tool in definitions.items():
+                # not sure if we should just pick MCP schema or make the schema something the user can pass
+                jsonschema.validate(instance=tool["metadata"], schema=MCP_TOOL_SCHEMA)
+                tools[name] = tool
+        except Exception as e:
+            # not sure if this should fail quietly, raise an error, or log it?
+            print(f"[tool-discovery] Failed to load tools from {self.module_name}: {e}")
```

Comment:
Instead of using `print`, we should make this a `self.log.error`. 

Before we do that, though, we'll need to make `ExtensionPoint` a `LoggingConfigurable` and set the `parent` trait to `ExtensionPackage`. I can work on this in a separate PR and we can rebase here once that's merged.

Reply:
See #1522. If/when that gets merged, we can rebase your PR to use the logger created there.